controller/tap/apiserver.go

```diff
@@ -0,0 +1,166 @@
+package tap
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/http"
+
+	"github.com/julienschmidt/httprouter"
+	pb "github.com/linkerd/linkerd2/controller/gen/controller/tap"
+	"github.com/linkerd/linkerd2/controller/k8s"
+	"github.com/linkerd/linkerd2/pkg/prometheus"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type apiServer struct {
+	router       *httprouter.Router
+	allowedNames []string
+	log          *logrus.Entry
+}
+
+// NewAPIServer creates a new server that implements the Tap APIService.
+func NewAPIServer(addr string, cert tls.Certificate, k8sAPI *k8s.API, client pb.TapClient, disableCommonNames bool) (*http.Server, net.Listener, error) {
+	clientCAPem, allowedNames, usernameHeader, groupHeader, err := apiServerAuth(k8sAPI)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// for development
+	if disableCommonNames {
+		allowedNames = []string{}
+	}
+
+	log := logrus.WithFields(logrus.Fields{
+		"component": "apiserver",
+		"addr":      addr,
+	})
+
+	h := &handler{
+		k8sAPI:         k8sAPI,
+		usernameHeader: usernameHeader,
+		groupHeader:    groupHeader,
+		tapClient:      client,
+		log:            log,
+	}
+
+	router := initRouter(h)
+
+	server := &apiServer{
+		router:       router,
+		allowedNames: allowedNames,
+		log:          log,
+	}
+
+	clientCertPool := x509.NewCertPool()
+	clientCertPool.AppendCertsFromPEM([]byte(clientCAPem))
+
+	wrappedServer := prometheus.WithTelemetry(server)
+
+	s := &http.Server{
+		Addr:    addr,
+		Handler: wrappedServer,
+		TLSConfig: &tls.Config{
+			Certificates: []tls.Certificate{cert},
+			ClientAuth:   tls.VerifyClientCertIfGiven,
+			ClientCAs:    clientCertPool,
+		},
+	}
+
+	lis, err := net.Listen("tcp", addr)
+	if err != nil {
+		log.Fatalf("net.Listen failed with: %s", err)
+	}
+
+	return s, lis, nil
+}
+
+// ServeHTTP handles all routes for the APIServer.
+func (a *apiServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	a.log.Debugf("ServeHTTP(): %+v", req)
+
+	// if `requestheader-allowed-names` was empty, allow any CN
```

Comment:
I'm gonna be pedantic about this comment here because this is potentially confusing.  in local development we won't be providing a client cert at all so there will be no client CN.  (and our TLS config doesn't require a client cert, it only validates one if it is provided)

Reply:
A bit more context, from https://kubernetes.io/docs/tasks/access-kubernetes-api/configure-aggregation-layer/#kubernetes-apiserver-client-authentication

```
The connection must be made using a client certificate whose CN is one of those listed in --requestheader-allowed-names. Note: You can set this option to blank as --requestheader-allowed-names="". This will indicate to an extension apiserver that any CN is acceptable.
```